lib/AST/FixByVal.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2021-present, Trail of Bits, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed in accordance with the terms specified in
+ * the LICENSE file found in the root directory of this source tree.
+ */
+
+#define GOOGLE_STRIP_LOG 1
+
+#include "rellic/AST/FixByVal.h"
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+#include <llvm/IR/IRBuilder.h>
+#include <llvm/Transforms/Utils/Cloning.h>
+
+namespace rellic {
+char FixByVal::ID = 0;
+
+FixByVal::FixByVal() : llvm::ModulePass(ID) {}
+
+void FixByVal::visitFunction(llvm::Function& old_func) {
+  auto module{old_func.getParent()};
+  std::vector<llvm::Type*> new_arg_types{};
+  auto has_byval{false};
+
+  for (auto& arg : old_func.args()) {
+    if (arg.hasByValAttr()) {
+      auto ptrtype{llvm::cast<llvm::PointerType>(arg.getType())};
+      new_arg_types.push_back(ptrtype->getElementType());
+      has_byval = true;
+    } else {
+      new_arg_types.push_back(arg.getType());
+    }
+  }
+
+  if (!has_byval) {
+    return;
+  }
+
+  auto addrspace{old_func.getAddressSpace()};
+  auto new_func_type{llvm::FunctionType::get(
+      old_func.getReturnType(), new_arg_types, old_func.isVarArg())};
+  auto new_func{llvm::Function::Create(new_func_type, old_func.getLinkage(),
+                                       addrspace, old_func.getName(), module)};
+  new_func->setSubprogram(old_func.getSubprogram());
+  auto bb{llvm::BasicBlock::Create(module->getContext(), "", new_func)};
+  llvm::IRBuilder<> builder{bb};
+  llvm::ValueToValueMapTy map{};
+
+  for (auto i{0U}; i < old_func.arg_size(); ++i) {
+    auto old_arg{old_func.getArg(i)};
+    auto new_arg{new_func->getArg(i)};
+    if (old_arg->hasByValAttr()) {
+      new_arg->setName(old_arg->getName());
+      auto ptrtype{llvm::cast<llvm::PointerType>(old_arg->getType())};
+      auto alloca_arg{builder.CreateAlloca(ptrtype->getElementType())};
+      builder.CreateStore(new_arg, alloca_arg);
+
+      map[old_arg] = alloca_arg;
+    } else {
+      new_arg->setName(old_arg->getName());
+      map[old_arg] = new_arg;
+    }
+  }
+  llvm::SmallVector<llvm::ReturnInst*> rets{};
+  llvm::CloneFunctionInto(new_func, &old_func, map, true, rets);
+  builder.CreateBr(new_func->getBasicBlockList().getNextNode(*bb));
+  subst[&old_func] = new_func;
+  old_func.removeFromParent();
+}
+
+void FixByVal::visitCallInst(llvm::CallInst& call) {
```

Comment:
I think we need to handle /all/ uses of the function, and replace them, not just calls. This is much trickier than it seems, though, because functions can be used inside of constants. You can probably copy some of the constant cloning/replacement code from remill [1]. It ends up being quite a bit of code, but I think we want to make sure we handle situations like function pointers in code and data.

[1] https://github.com/lifting-bits/remill/blob/master/lib/BC/Util.cpp#L1576-L1616

Reply:
It is probably more complicated than what my the cloning code does; you'll need to decide/choose between:
 - if there's a function pointer in a constant, then do we cast the new version of the function to the old type; or
 - do we re-type the constant/global, and then find the effects of that.

I'm realizing that an llvm transform is opening us up to some really tricky issues. Maybe one possibility is this, that side-steps our things:
 - the original is retained, but changed so that it does a load of the `byval` pointer and calls the new version.

Reply:
Or, we fall back to modifying how `byval`-attributed parameters are handled in Rellic.

Reply:
I guess the latter was the easiest one :)